driver-core/src/main/com/mongodb/ClientBulkWriteException.java

```diff
@@ -0,0 +1,138 @@
+        this.writeErrors = writeErrors == null ? emptyMap() : unmodifiableMap(writeErrors);
+        this.partialResult = partialResult;
+    }
+
+    private static String message(
+            @Nullable final MongoException error,
+            @Nullable final List<WriteConcernError> writeConcernErrors,
+            @Nullable final Map<Integer, WriteError> writeErrors,
+            @Nullable final ClientBulkWriteResult partialResult,
+            final ServerAddress serverAddress) {
+        return "Client-level bulk write operation error on server " + serverAddress + "."
+                + (error == null ? "" : " Top-level error: " + error + ".")
+                + (writeErrors == null || writeErrors.isEmpty() ? "" : " Write errors: " + writeErrors + ".")
+                + (writeConcernErrors == null || writeConcernErrors.isEmpty() ? "" : " Write concern errors: " + writeConcernErrors + ".")
+                + (partialResult == null ? "" : " Partial result: " + partialResult + ".");
+    }
+
+    /**
+     * The top-level error. That is an error that is neither a {@linkplain #getWriteConcernErrors() write concern error},
+     * nor is an {@linkplain #getWriteErrors() error of an individual write operation}.
+     *
+     * @return The top-level error. {@linkplain Optional#isPresent() Present} only if a top-level error occurred.
+     */
+    public Optional<MongoException> getError() {
+        return ofNullable(error);
+    }
+
+    /**
+     * The {@link WriteConcernError}s that occurred while executing the client-level bulk write operation.
+     * <p>
+     * There are no guarantees on mutability of the {@link List} returned.</p>
+     *
+     * @return The {@link WriteConcernError}s.
+     */
+    public List<WriteConcernError> getWriteConcernErrors() {
+        return writeConcernErrors;
+    }
+
+    /**
+     * The indexed {@link WriteError}s.
+     * The {@linkplain Map#keySet() keys} are the indexes of the corresponding {@link ClientNamespacedWriteModel}s
+     * in the corresponding client-level bulk write operation.
+     * <p>
+     * There are no guarantees on mutability or iteration order of the {@link Map} returned.</p>
+     *
+     * @return The indexed {@link WriteError}s.
+     * @see ClientBulkWriteResult.Verbose#getInsertResults()
+     * @see ClientBulkWriteResult.Verbose#getUpdateResults()
+     * @see ClientBulkWriteResult.Verbose#getDeleteResults()
+     */
+    public Map<Integer, WriteError> getWriteErrors() {
+        return writeErrors;
+    }
+
+    /**
+     * The result of the successful part of a client-level bulk write operation.
+     *
+     * @return The successful partial result. {@linkplain Optional#isPresent() Present} only if at least one
+     * {@linkplain ClientNamespacedWriteModel individual write operation} succeed.
+     */
+    public Optional<ClientBulkWriteResult> getPartialResult() {
```

Comment:
`BULK-TODO` Should not be `Optional`, see https://jira.mongodb.org/browse/DRIVERS-2975.

Reply:
Judging from https://mongodb.slack.com/archives/C035ZJL6CQN/p1726775538071889?thread_ts=1726074791.410459&cid=C035ZJL6CQN, the result will stay optional.